### ISSUE TITLE
Add Java integration tests to PR CI builds

### ISSUE THREAD

**Issue by benwtrent** [elastic/ml-cpp#766]
It would be useful to catch CI failures between the ML Plugin and the C++ repo early. This will allow us to catch issues before the artifact is published. 

I think the test process would be something akin to:

- Clone ES repository
- Run the following
```
./gradlew :x-pack:plugin:integtest -Dtests.class=org.elasticsearch.xpack.test.rest.XPackRestIT -Dtests.method="test {p0=ml/*}"

./gradlew :x-pack:plugin:ml:qa:native-multi-node-tests:integTest
```
And the `30_ml_jobs_crud` `90_ml_data_frame_analytics_crud` rolling upgrade tests.

**Comment by mark-vieira** [comment]
To clarify though, we want to run those tests using the ML code in the PR. So we'd so something like:

1. Build the ML code
2. Publish it to some local repo
3. Checkout the ES repo
4. Configure it such that it uses the locally published ML artifact
5. Run some subset of the ES test suite specific to ML

The details in the middle there about how we pass around the ML artifact can be tweaked but that's the basic idea. The key point being that instead of running the ES tests using the most recently published ML snapshot artifact, we use the locally built one.

**Comment by droberts195** [comment]
@mark-vieira do you know if it’s possible for an Ivy repo configured in a `build.gradle` to be a `file:///` URL? If it is I think that will be the easiest way to get the ES build to pick up the C++ build from the PR.

**Comment by mark-vieira** [comment]
Yes, absolutely. Maven/Ivy repos with a `file://` scheme function in exactly the same way as remote repos with using HTTP.

**Comment by atorok** [comment]
Before we move to implementation I think it's worth to discuss weather we still want to keep a separate repo. 
We could have dedicated CI jobs and PR checks ( based on the ml label ) initially, and gradually integrate with the build better so that  the ML team is better able to make use of the work going into build and tooling ( e.x. build cache,  build scans etc. )

**Comment by mark-vieira** [comment]
For what it's worth, this is a great example of what I brought up in [this thread](https://groups.google.com/a/elastic.co/forum/?utm_medium=email&utm_source=footer#!msg/es-core-infra/Q2bOmaeXQ2c/M291Po2DCwAJ) about issue naming and distilling down the fundamental goal of some initiative.

While "Add Java integration tests to PR CI builds" is one way of phrasing what we want to achieve, I think the goal here is higher level than this. What we want to do is increase the level of test coverage on pull requests to this repository such that PRs serve as a sufficient quality gate for merge acceptance. Right now this is not the case as changes to code in this repo can cause issues when integrating with the upstream elasticsearch code. With that in mind, something like "ML pull requests are integration tested with upstream Elasticsearch". Ben also summarizes this goal well in the issue description with:

> It would be useful to catch CI failures between the ML Plugin and the C++ repo early.

As @atorok mentions, another way to achieve this, and other benefits, would be to fold this into the `elastic/elasticsearch` repo. It's certainly worth a discussion, but is a considerably more considerable and disruptive change as there are many development, test and release processes that function based on the existing topology.

In consideration of this discussion, I'm personally most interested in insights from the ML developers themselves as this would primarily affect their day to day work. Similar to the motivation to folk in x-pack, we'd have to consider how this would alleviate developer pain. Simply, what are the day to day frustrations and pain points with developing ML in the current multi-repo configuration (this test integration issue amongst them)? How might folding this repository into the existing Elasticsearch code base and integrating their build processes solve those issues? Considering that method is a non-trivial undertaking I'm hesitant to consider it without significant motivation, and primarily from the folks that would be most directly impacted by such a change. Thoughts?

**Comment by droberts195** [comment]
I don't think we should merge the repos.  There is immense complexity in doing that.  For example:

1. The Elasticsearch build assumes that whatever platform it's running on it will be able to build the distributions for Mac, Windows and Linux.  Yet to build the ML C++ code for Windows you need a Windows VM.  We cannot create a requirement that every Elasticsearch developer has to have a suitably configured Windows VM just to be able to successfully run `./gradlew assemble`.
2. Elasticsearch Java developers don't want to sit and wait 20 minutes for the ML C++ to compile when they pull latest master and try to run some Java unit tests locally just because an ML C++ developer happened to commit a change to some widely used header file immediately before.
3. A developer working on a Mac trying to debug a complex problem in ML wants to build the ML C++ natively on their Mac so the debug symbols are accessible by the debugger.  But another developer on a Mac doesn't even have clang installed and doesn't want to know the details of how to build and debug the ML C++.

None of these problems are impossible to solve, but they would be very time consuming to solve with disruption to the whole ES dev team while all the edge cases are solved, and we have more important things to work on.

I have opened elastic/elasticsearch#48702 to make the URL for the Ivy repo where the ML C++ artifact is downloaded from configurable in the ES build.  If that PR is acceptable then I can make the remaining changes by editing the `dev-tools/jenkins_ci.sh` script in this repo.  This will provide a very low cost way to make our PR CI builds on Linux run a subset of ES integration tests using the C++ artifact built by that particular CI build.  By going in this direction we can improve the situation with C++ changes accidentally breaking Java tests with ~1 day of dev effort rather than weeks/months to merge the repos and solve all the resulting problems.

**Comment by droberts195** [comment]
> And the `30_ml_jobs_crud` `90_ml_data_frame_analytics_crud` rolling upgrade tests.

This introduces further complexities around determining the correct BWC branches and commits to use.  In the first instance I will try to get this working for our basic YAML tests and multi-node QA tests.  Once that has been working for a few weeks we can assess adding BWC tests.

**Comment by atorok** [comment]
The logic in the ES build around bwc should be applicable, especially for PR check where we just take the latest off of snapshot branches, but we do need a way to build the ml-cpp binaries on those if we want to be authentic. 

Running only the ML part of bwc will add complexity too. 
We should consider adding their own bwc test. I know there is no precedent for that, 
and we would need to add a plugin for the bwc setup to avoid repeating it, in yet another plugin, but still worth considering I think. 
It might help even outside of this context as in the bwc jobs  we generally have idle threads, so there is room for running more things in parallel.

@droberts195  An additional clarification: when you say we should not merge the repos, do you mean in a general way or that we shouldn't condition this issue on it ? 
I'm still inclined to think that we would be better off long-term. I fully agree that if there's a faster way to get some/all  PR checks running we should definitely do it.

**Comment by droberts195** [comment]
> An additional clarification: when you say we should not merge the repos, do you mean in a general way or that we shouldn't condition this issue on it ?

I meant in a general way.  I am not convinced we will be better off in the longer term by merging the repos.  A couple of weeks ago some types removal work in Elasticsearch broke Kibana.  It could have been avoided if the `elasticsearch` and `kibana` repos were merged, and every PR ran both the Elasticsearch and Kibana tests.  But I think most people would think that was an overreaction to a fairly rare problem.  If we went in that direction then the logical conclusion would be that all repos associated with the Elastic stack would be merged into one.  There are obviously benefits of merging repos but also drawbacks.  I don't think the benefits far outweigh the drawbacks, so we might as well leave things as they are.
